File: libsrm_cpp/src/srm.cpp
/**
 * Implementation source file for the Statistical Region Merging Algorithm
 */
#include <cmath>
#include <string>
#include <sstream>
#include <map>

#include "../include/libsrm_cpp/srm.hpp"
#include "../include/libsrm_cpp/utils.hpp"

#include <opencv2/opencv.hpp>


/**
 * SRM: The constructor for the Algorithm Object
 * @param _Q: by default Q = 32, and it guides / dictates
 *            the granularity of the segmentation
 * @param _img_size: the size of the image (width * height)
 * @param _max_regions: The maximum ammount of regions chosen
 *                      for a given picture (to prevent overmerging)
 * @param _min_size: parameter chosen to avoid undermerging
 */
SRM::SRM(int _Q, int _max_regions, double _min_size, cv::Mat& image) : img(image) {
    this->imgRows = this->img.rows;
    this->imgCols = this->img.cols;
    this->Q = _Q;
    this->g = 256;
    this->imgSize = this->imgRows * this->imgCols;
    this->maxRegions = _max_regions;
    this->minSize = _min_size * static_cast<float>(this->imgSize);
    this->delta = std::log(6.0) + 2.0 * std::log(static_cast<double>(this->imgSize));

    this->rank = std::vector<int>(this->imgSize, 1);

    for (int i = 0; i < this->imgSize; i++) {
        this->parent.push_back(i);
    }

    if (this->maxRegions < 0) {
        this->maxRegions = this->imgSize;
    }
}


/**
 * run:
 * @param path: Path to the Image
 */
void SRM::run() {
    
    cv::Mat floatImage;
    this->img.convertTo(floatImage, CV_32F);
    this->img = floatImage.reshape(1, this->getImageSize());


    // 1) Sort the list of pairs (4-Connected Neighbors)
    std::vector<std::pair<int, int>> edgeList = this->makeEdgePairList(this->getImageRows(), this->getImageCols());


    // 2) Iterate through the list of pairs and perform the statistical test
    for (const std::pair<int, int>& p : edgeList) {
        
        int parentA = this->getParent(p.first);
        int parentB = this->getParent(p.second);

        if ((parentA != parentB) && (this->predicate(parentA, parentB))) {
            this->merge(parentA, parentB);
        } 

    }

    // // 3) Check for occluded regions
    // if (this->maxRegions > 0) {
    //     this->mergeOcclusions();
    // }

    // // 4) Check for smaller regions
    // if (this->minSize > 0) {
    //     this->mergeSmallerRegions();
    // }

    // 5) Change the color of the image accordingly (actual segmentation)
    for (int i = 0; i < this->imgSize; i++) {
        cv::Vec3f color = this->img.at<cv::Vec3f>(0, this->getParent(i));
        this->img.at<cv::Vec3f>(0, i) = color;
    }

    // 6) Reshape the image and save it to the object's state
    this->img = this->img.reshape(3, this->getImageRows());
    this->img.convertTo(this->img, CV_8U);

}


/**
 * makeEdgePairList: 
 * 
 */
std::vector<std::pair<int, int>>& SRM::makeEdgePairList(int rows, int cols) {

    std::vector<std::pair<int, int>>* edgeList = new std::vector<std::pair<int, int>>;

    int index = 0;
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            index = i * rows + j;
            if (i != rows - 1) {
                edgeList->push_back(std::make_pair(index, index + cols));
            }
            if (j != cols - 1) {
                edgeList->push_back(std::make_pair(index, index + 1));
            }
        }
    }

    this->sortEdgePairs(*edgeList);

    return *edgeList;
}


/**
 * sortEdgePairs: 
 * 
 */
void SRM::sortEdgePairs(std::vector<std::pair<int,int>>& edgeList) {
    
        DiffComparator diffComparator(this->img);

        std::stable_sort(edgeList.begin(), edgeList.end(),
            [&diffComparator](const std::pair<int, int>& a, const std::pair<int, int>& b) {
                return diffComparator(a) < diffComparator(b);
            });

}


/**
 * evaluatePredicate: The P(.) function from the paper
 * @param parentA: The first region of interest
 * @param parentB: The second region of interest
 */
bool SRM::predicate(int parentA, int parentB) {

    double predicateA = this->evaluatePredicate(parentA);
    double predicateB = this->evaluatePredicate(parentB);
    cv::Vec3f comparrison = (this->img.at<cv::Vec3f>(0, parentA) - this->img.at<cv::Vec3f>(0, parentB));
    comparrison = comparrison.mul(comparrison);
    
    auto checkPredicate = [](const cv::Vec3f& squaredDiff, double threshold) {
        return (squaredDiff[0] < threshold) && (squaredDiff[1] < threshold) && (squaredDiff[2] < threshold);
    };

    return (checkPredicate(comparrison, (predicateA + predicateB)));
}

/**
 * predicate: Computes the actual predicate
 *            on a single region
 */
double SRM::evaluatePredicate(int parentA) {

    return (
        (std::pow(this->g, 2) / static_cast<double>(2 * this->Q * this->rank[parentA])) *
        static_cast<double>(std::min(this->g, this->rank[parentA])) *
        std::log(static_cast<double>(this->rank[parentA]) + 1) +
        this->delta
    );
}

/**
 * merge: Performs the actual merging of the regions
 * @param parentA: Region 1
 * @param parentB: Region 2
 * 
 */
void SRM::merge(int parentA, int parentB) {
    int sum1 = this->rank[parentA];
    int sum2 = this->rank[parentB];
    
    cv::Vec3f color = (
        (this->img.at<cv::Vec3f>(0, parentA) * sum1) + 
        (this->img.at<cv::Vec3f>(0, parentB) * sum2)
    ) / 
    (static_cast<float>(sum1) + static_cast<float>(sum2));

    if (sum1 < sum2) {
        int swap = parentA;
        parentA = parentB;
        parentB = parentA;
    }

    this->parent[parentB] = parentA;
    this->rank[parentA] += this->rank[parentB];
    this->img.at<cv::Vec3f>(0, parentA) = color;
}


/**
 * getParent: Obtains the root element in the UnionFind
 * @param parentA: Query region
 * 
 */
int SRM::getParent(int parentA) {
    if (this->parent[parentA] == parentA) {
        return parentA;
    }
    // Lazy propagation
    int p = this->getParent(this->parent[parentA]);
    return p;
}

/**
 * mergeOcclusions:
 * 
 */
void SRM::mergeOcclusions() {
    for (int i = 1; i < this->getImageSize(); i++) {
        int r1 = this->getParent(i);
        int r2 = this->getParent(i - 1);
        if ((r1 != r2) && (this->rank[r1] <= this->minSize)) {
            this->merge(r1, r2);
        }
    }
}

/**
 * mergeSmallerRegions: 
 */
void SRM::mergeSmallerRegions() {
    
    std::vector<int> maxParents = this->retrieveMaxParents();
    if (maxParents.size() > this->maxRegions) {
        std::vector<int> allowedParents(maxParents.begin(), maxParents.begin() + this->maxRegions);
        for (int i = 0; i < this->getImageSize(); i++) {
            int r1 = this->getParent(i);
            int r2 = this->getParent(i - 1);
            if ((r1 != r2) && (std::find(allowedParents.begin(), allowedParents.end(), r1) == allowedParents.end())) {
                this->merge(r1, r2);
            }
        }
    }

}


/**
 * retrieveMaxParents:
 * 
 */
std::vector<int>& SRM::retrieveMaxParents() {

    std::vector<int>* sortedKeys = new std::vector<int>;
    std::map<int, int> parentsCounts;

    for (int pt : this->parent) {
        parentsCounts[pt]++;
    }

    std::vector<std::pair<int, int>> countsVector(parentsCounts.begin(), parentsCounts.end());

    std::sort(countsVector.begin(), countsVector.end(),
              [](const std::pair<int, int>& a, const std::pair<int, int>& b) {
                  return a.second > b.second;
              });

    for (const auto& pair : countsVector) {
        sortedKeys->push_back(pair.first);
    }

    return *sortedKeys;
}

/**
 * Saves the segmented image
 */
void SRM::saveImage(const std::string& asbPath) {

    std::ostringstream oss;
    oss << "Segmented_" << getBaseName(asbPath);
    std::string fileName = oss.str();

    cv::imwrite(fileName, this->img);

}

int SRM::getImageRows() const {
    return this->imgRows;
}

int SRM::getImageCols() const {
    return this->imgCols;
}

int SRM::getImageSize() const {
    return this->imgSize;
}

File: libsrm_rust/src/algorithms/srm.rs
/**
* Implementation of the Statistical Region Merging Algorithm
*/

use opencv::{
    core::Mat,
    core::CV_32F,
    core::CV_8U,
    core::Vec3f,
    core::VecN,
    core::Vector,
    prelude::MatTrait,
    prelude::MatTraitConst,
    imgcodecs::imwrite,
};

use std::path::Path;


pub struct SRM<'a> {
    q: i32,
    g: i32,
    delta: f32,
    img_size: i32,
    min_size: f32,
    max_regions: i32,
    img_rows: i32,
    img_cols: i32,
    rank: Vec<i32>,
    parent: Vec<i32>,
    img: &'a Mat,
    reshaped_image: Mat,
}



impl<'a> SRM<'a> {

    pub fn new(q: i32, max_regions: i32, min_size: f32, img: &'a Mat) -> Result<Self, opencv::Error> {

        let img_size: i32 = img.rows() * img.cols();
        let delta: f32 = f32::ln(6.0) + 2.0 * f32::ln(img_size as f32);
        let min_size: f32 = min_size * img_size as f32;

        Ok(
            SRM {
                q,
                g: 256,
                delta,
                img_size,
                min_size,
                max_regions,
                img_rows: img.rows(),
                img_cols: img.cols(),
                rank: vec![1; img_size as usize],
                parent: (0..img_size as i32).collect(),
                img,
                reshaped_image: Mat::default(),
            }
        )
    }


    pub fn run(&mut self) -> Result<(), Box<dyn std::error::Error>> {

        let mut float_image = Mat::default();
        self.img.convert_to(&mut float_image, CV_32F, 1.0, 0.0)?;
        let reshaped_img = float_image.reshape_mut(3, self.img_size)?.try_clone()?;
        self.reshaped_image = reshaped_img;


        // 1) Sort the list of pairs (4-Connected Neighbors)
        let edge_list: Vec<(i32, i32)> = self.make_edge_pair_list();

        // // 2) Iterate through the list of pairs and perform the statistical test
        for p in &edge_list {
            let parent_a = self.get_parent(p.0);
            let parent_b = self.get_parent(p.1);

            if parent_a != parent_b && self.predicate(parent_a, parent_b) {
                let _ = self.merge(parent_a, parent_b);
            }
        }

        // 3) Merge Occlusions if present
        if self.max_regions > 0 {
            self.merge_occlusions();
        }

        // 4) Prepare the segmented image's colors
        let _ = self.process_image();

        // 5) Reshape and re-convert the image
        let original_shape_image = self.reshaped_image
        .reshape(3, self.img_rows)?
        .try_clone()?;
        original_shape_image.convert_to(&mut self.reshaped_image, CV_8U, 1.0, 0.0)?;


        Ok(())

    }

    pub fn make_edge_pair_list(&self) -> Vec<(i32, i32)> {

        let mut edge_list = Vec::new();
        let rows = self.img_rows;
        let cols = self.img_cols;

        for i in 0..rows {
            for j in 0..cols {
                let index = i * rows + j;
                // Add vertical edge (index -> index + cols)
                if i != rows - 1 {
                    edge_list.push((index, index + cols));
                }
                // Add horizontal edge (index -> index + 1)
                if j != cols - 1 {
                    edge_list.push((index, index + 1));
                }
            }
        }

        self.sort_edge_pairs(&mut edge_list);

        edge_list
    }

    pub fn sort_edge_pairs(&self, edge_pairs: &mut Vec<(i32, i32)>) {
        edge_pairs.sort_by(|a, b| {
            a.0.cmp(&b.0).then_with(|| a.1.cmp(&b.1))
        });
    }

    pub fn evaluate_predicate(&self, parent_a: i32) -> f32 {
        ((self.g.pow(2) as f32) / (2.0 * self.q as f32 * self.rank[parent_a as usize] as f32))
        * (self.g as f32).min(self.rank[parent_a as usize] as f32)
        * ((self.rank[parent_a as usize] as f32) + 1.0).ln()
        + self.delta
    }


    pub fn predicate(&mut self, parent_a: i32, parent_b: i32) -> bool {

        let predicate_a = self.evaluate_predicate(parent_a);
        let predicate_b = self.evaluate_predicate(parent_b);


        let pixel_a = self.reshaped_image.at::<Vec3f>(parent_a as i32).unwrap();
        let pixel_b = self.reshaped_image.at::<Vec3f>(parent_b as i32).unwrap();

        // Computes the squared difference using the .mul() method (quicker)
        let mut comparison = self.subtract_pixels(pixel_a, pixel_b);
        comparison = comparison.mul(comparison);

        self.check_predicate(comparison, predicate_a + predicate_b)

    }

    fn subtract_pixels(
        &self,
        pixel_a: &VecN<f32, 3>,
        pixel_b: &VecN<f32, 3>,
    ) -> VecN<f32, 3> {

        VecN::<f32, 3>::from([
            pixel_a[0] - pixel_b[0],
            pixel_a[1] - pixel_b[1],
            pixel_a[2] - pixel_b[2],
        ])

    }

    fn check_predicate(
        &self,
        squared_diff: Vec3f,
        threshold: f32
    ) -> bool {
        squared_diff[0] < threshold as f32 && squared_diff[1] < threshold as f32 && squared_diff[2] < threshold as f32
    }

    pub fn merge(&mut self, parent_a: i32, parent_b: i32) -> Result<(), Box<dyn std::error::Error>> {


        let sum1 = self.rank[parent_a as usize];
        let sum2 = self.rank[parent_b as usize];


        let pixel_a = *self.reshaped_image.at::<Vec3f>(parent_a)?;
        let pixel_b = *self.reshaped_image.at::<Vec3f>(parent_b)?;


        let color_result = (pixel_a * (sum1 as f32) + pixel_b * (sum2 as f32)) / ((sum1 + sum2) as f32);

        let (parent_a, parent_b) = if sum1 < sum2 {
            (parent_b, parent_a)
        } else {
            (parent_a, parent_b)
        };


        self.parent[parent_b as usize] = parent_a;
        self.rank[parent_a as usize] += self.rank[parent_b as usize];

        *self.reshaped_image.at_mut::<Vec3f>(parent_a)? = color_result;

        Ok(())

    }

    pub fn get_parent(&mut self, parent_a: i32) -> i32 {

        if self.parent[parent_a as usize] == parent_a {
            return parent_a;
        }
        // Lazy propagation
        let p = self.get_parent(self.parent[parent_a as usize]);
        self.parent[parent_a as usize] = p; // Path compression
        p

    }

    pub fn merge_occlusions(&mut self) {

        for i in 1..self.img_size {
            let r1 = self.get_parent(i);
            let r2 = self.get_parent(i - 1);
            if r1 != r2 && (self.rank[r1 as usize] as f32) <= self.min_size {
                let _ = self.merge(r1, r2);
            }
        }

    }

    pub fn process_image(&mut self) -> Result<(), Box<dyn std::error::Error>> {


        for i in 0..self.img_size {

            let parent_idx = self.get_parent(i);
            let color = self.reshaped_image.at::<Vec3f>(parent_idx)?;

            *self.reshaped_image.at_mut::<Vec3f>(i)? = *color;
        }

        Ok(())
    }


    pub fn save_image(&self, abs_path: &str) -> Result<(), Box<dyn std::error::Error>> {

        let base_name = Path::new(abs_path)
        .file_name()
        .and_then(|name| name.to_str())
        .ok_or("Failed to extract base name")?;

        let file_name = format!("Segmented-{}", base_name);
        let params = Vector::<i32>::new();

        imwrite(&file_name, &self.reshaped_image, &params)?;


        Ok(())
    }

}
